model_defs/attention.py

```diff
@@ -0,0 +1,257 @@
+
+        self.phi = conv_nd(in_channels=self.in_channels, out_channels=self.inter_channels,
+                           kernel_size=1, stride=1, padding=0)
+        nn.init.kaiming_uniform(self.phi.weight, a=1)
+        nn.init.constant(self.phi.bias, 0)
+
+        self.combination_mode = combination_mode
+        if combination_mode == "stack-reduce":
+            self.output_reduce = conv_nd(in_channels=2*self.in_channels,
+                                         out_channels=self.in_channels,
+                                         kernel_size=1,
+                                         stride=1,
+                                         padding=0)
+
+    def forward(self, x, target, mode="average"):
+        '''
+        :param x: B x C x H x W
+        :param y: B x C x h x w
+        :return:
+        '''
+
+        assert mode in ["softmax", "average"]
+
+        batch_size = x.size(0)
+
+        # embed the object features
+        g_target = self.g(target).view(batch_size, self.inter_channels, -1)
+        # B x hw x C
+        g_target = g_target.permute(0, 2, 1)
+
+        # B x C x HW
+        theta_x = self.theta(x).view(batch_size, self.inter_channels, -1)
+        # B x HW x C
+        theta_x = theta_x.permute(0, 2, 1)
+        # B x C x hw
+        phi_target = self.phi(target).view(batch_size, self.inter_channels, -1)
+
+        # B x HW x hw
+        f = torch.matmul(theta_x, phi_target)
+        if mode == "softmax":
+            f_div_C = F.softmax(f, dim=-1)
+        else:
+            f_div_C = f / f.size(-1)
+
+        # B x HW x C
+        y = torch.matmul(f_div_C, g_target)
+
+        get_norm = lambda z, dim: torch.norm(z, p=2, dim=dim)
+        print("="*10)
+        print("Input x norm: {}".format(
+            get_norm(x, 1).mean().data.cpu().numpy()))
+        print("Target norm: {}".format(
+            get_norm(target, 1).mean().data.cpu().numpy()))
+        print("Embedded target norm: {}/{}/{}".format(
+            get_norm(g_target, 2).min().data.cpu().numpy(),
+            get_norm(g_target, 2).max().data.cpu().numpy(),
+            get_norm(g_target, 2).mean().data.cpu().numpy()))
+        print("Min/Max/Mean abs correlation: {}/{}/{}".format(
+            f.abs().min().data.cpu().numpy(),
+            f.abs().max().data.cpu().numpy(),
+            f.abs().mean().data.cpu().numpy()))
+        print("Aggregated norm: {}/{}/{}".format(
+            get_norm(y, 2).min().data.cpu().numpy(),
+            get_norm(y, 2).max().data.cpu().numpy(),
+            get_norm(y, 2).mean().data.cpu().numpy()))
+
+        # B x C x HW
+        y = y.permute(0, 2, 1).contiguous()
+        y = y.view(batch_size, self.inter_channels, *x.size()[2:])
+        W_y = self.W(y)
+        if self.combination_mode == "stack-reduce":
+            z = torch.cat([W_y, x], dim=1)
+            z = self.output_reduce(z)
+        elif self.combination_mode == "attention-only":
+            z = W_y
```

Comment:
Given that in TDID final version image features do not contribute much, I think it's good to try and train this version with "average" aggregation